### ISSUE TITLE
Heavy CPU usage and possible endless Close() loop when using pgx pool

### ISSUE THREAD

**Issue by rupor-github** [jackc/pgx#800]
Hello,

Context: **ubuntu 16.04, go1.13.4 or go1.14.2, pgx 4.7.2, postgres 9.6.16**.

I have simple set of worker goroutines executing single pgxpool.Query with SELECT statement. Context deadline is set for 1 second, but under normal conditions it is much faster. Basically code looks like this (simplified):

```
for {
    rows, err := pgxpool.Query()
    if err != nil {
        report and exit
    }
    defer rows.Close()
    for rows.Next() {
        err = rows.Scan(...)
        if err != nil {
             report and exit
        }
    }
    if err := rows.Err(); err != nil {
             report and exit
    }
    ....
)
```
One goroutine per connection, program runs perfectly for days under production load until heavy replication arrives to postgres servers it is connected to. When this happens relatively small number of recovery errors is being reported by servers (~200 errors with production load ~15000 queries per second on 20 connections in the pool):

`ERROR: canceling statement due to conflict with recovery (SQLSTATE 40001)` and
`FATAL: terminating connection due to conflict with recovery (SQLSTATE 40001)`

After that program starts consuming a lot of CPU executing endless Close(). From what I could tell all gorourines are getting connection and attempting to close it, but all connections which have been already closed by postgress. tcpdump shows no communication between program and postgres, no connections are reported by postgres and no errors are being returned by pgxpool.Query call.

For all workers stacks look like this one (I selected the deepest one):
```
(dlv) stack
 0  0x0000000000685701 in net.(*conn).Read
    at net/net.go:186
 1  0x00000000004ae8c7 in io.ReadAtLeast
    at io/io.go:310
 2  0x0000000000a1c9e2 in github.com/jackc/chunkreader/v2.(*ChunkReader).appendAtLeast
    at github.com/jackc/chunkreader/v2@v2.0.1/chunkreader.go:88
 3  0x0000000000a1c9e2 in github.com/jackc/chunkreader/v2.(*ChunkReader).Next
    at github.com/jackc/chunkreader/v2@v2.0.1/chunkreader.go:78
 4  0x0000000000a264e7 in github.com/jackc/pgproto3/v2.(*Frontend).Receive
    at github.com/jackc/pgproto3/v2@v2.0.2/frontend.go:71
 5  0x0000000000a68b74 in github.com/jackc/pgconn.(*PgConn).receiveMessage
    at github.com/jackc/pgconn@v1.6.2/pgconn.go:443
 6  0x0000000000a70585 in github.com/jackc/pgconn.(*ResultReader).receiveMessage
    at github.com/jackc/pgconn@v1.6.2/pgconn.go:1407
 7  0x0000000000a6f8ce in github.com/jackc/pgconn.(*ResultReader).Close
    at github.com/jackc/pgconn@v1.6.2/pgconn.go:1385
 8  0x0000000000b315f8 in github.com/jackc/pgx/v4.(*connRows).Close
    at github.com/jackc/pgx/v4@v4.7.2/rows.go:130
 9  0x0000000000b3178f in github.com/jackc/pgx/v4.(*connRows).Next
    at github.com/jackc/pgx/v4@v4.7.2/rows.go:184
10  0x0000000000b3e98a in github.com/jackc/pgx/v4/pgxpool.(*poolRows).Next
    at github.com/jackc/pgx/v4@v4.7.2/pgxpool/rows.go:62
11  ...
12  ...
13  start of the worker goroutine
```

Here is how goroutine analysis for pprof trace looks like in this case (20 workers for 20 connections in the pool):

![image](https://user-images.githubusercontent.com/5368784/88812588-e4573f00-d185-11ea-8592-41e9eceaf84e.png)

Normally however the picture would be quite different:

![image](https://user-images.githubusercontent.com/5368784/88813024-6b0c1c00-d186-11ea-9575-ae0b25aa2050.png)

For me it looks like dead connections are stack in the pool - looking at the code I cannot figure out at what point error was not reported and handled properly...

**Comment by jackc** [comment]
I would expect PostgreSQL to close the TCP connection after it sends the fatal error. Then I would expect the `net.(*conn).Read` would immediately fail.

It appears one of those assumptions is faulty. Is this connected directly to PostgreSQL or through a load balancer?

Do these fatal errors always trigger this cascading failure or are they handled properly sometimes?

Also, that stack trace looks like it is stuck while reading the results of a query, not that it is continually reusing dead connections.

Does the system ever recover? The current execution path after receiving a fatal error goes through `pgconn.(*PgConn).asyncClose()` (https://github.com/jackc/pgconn/blob/37c9edc242e83750fcfbef327001fd65603d63d0/pgconn.go#L534). If PostgreSQL doesn't close the TCP connection then pgx will. That has a worst case time of 15 seconds. It tries to do an orderly shutdown of the connection. It might make more sense to close the net.Conn directly without sending the terminate message in case of receiving a fatal error but it shouldn't permanently block as it is now.

But if the system never recovers that seems to indicate the problem lies elsewhere.

Oh, and I just saw you do have a context deadline set. Regardless of anything else that should unblock the reads and mark the connection as dead.

The only other thing that jumped out at me is the `defer rows.Close()`. I'm guessing your real code is different, but as it is written in the example those defers will accumulate indefinitely as they execute on function exit not loop block exit.

**Comment by rupor-github** [comment]
1. It is connected to postges directly - no balancers.
2. It looks like it is reading results from a query and attempts to close in Next(), however postgres reports no live connections and tcpdump showns no traffic in and out of postgres. At the same time I have gouroutines which are doing stance like this one while eating 80% of my 28 CPUs:
![image](https://user-images.githubusercontent.com/5368784/88865132-fc09e400-d1d4-11ea-82ba-468c1a18bab8.png)
I am not sure where it went wrong - read pgx code, but was not able to figure it out. AcquiredCons are 0 and TotalCons are maxed out and it stays this way. To me it looks like pgxpool thinks that connections in the pool are valid and is not destroying them while postgres thinks that it closed them down. Lost sync with wire protocol? Missed error condition (one of the Close() routines in stack trace actually has for{} which will be exited only when connection is ready for query or error arrives)? At any rate - it is not blocked, it is actively executing Close(). A lot of them on all cores :)
3. defers should be fine - I took code out of simple function to simplify 😄 
4. System never recovers - whatever left of CPU -- ~20% is running from the previously filled cache indefinitely.

I was analyzing my code for a while - as I mentioned before, program is working under heavy load for days and weeks until recovery error arrives from Postgres. This is difficult to reproduce (I cannot create a small piece of code - need production load) but it is 100% reliable in 5 data centers around the globe.

At the moment I am trying to run with my own pool - it is pretty simple, so I just open connections statically trying to see if I could get a better trace. Also it never happened with previous version of pgx - v3 is surviving those postgres errors without a problem.

**Comment by jackc** [comment]
I think I have managed to duplicate this. Or at least something similar that eventually locks while using a lot of CPU in under 30 seconds. Still trying to narrow the reproduction case and trace the underlying problem.

```go
func TestPoolQueryConnTerminated(t *testing.T) {
	poolSize := 10
	workerCount := 20
	jobsPerWorker := 500

	config, err := pgxpool.ParseConfig(os.Getenv("PGX_TEST_DATABASE"))
	require.NoError(t, err)
	config.MaxConns = int32(poolSize)

	db, err := pgxpool.ConnectConfig(context.Background(), config)
	require.NoError(t, err)
	defer db.Close()

	doneChan := make(chan struct{})

	go func() {
		ticker := time.NewTicker(500 * time.Millisecond)
		for {
			select {
			case <-ticker.C:
				stat := db.Stat()
				fmt.Printf("Acquired: %v, Idle: %v, Total: %v\n", stat.AcquiredConns(), stat.IdleConns(), stat.TotalConns())
			case <-doneChan:
				ticker.Stop()
				return
			}
		}
	}()

	var wg sync.WaitGroup
	for i := 0; i < workerCount; i++ {
		wg.Add(1)
		go func() {
			for i := 0; i < jobsPerWorker; i++ {
				sql := `select n,
  case when random() > 0.99 then
    pg_terminate_backend((
      select pid
      from pg_stat_activity
      where datname = current_database()
        and usename = current_user
      order by random()
      limit 1
    ))
  end
from generate_series(1, 200) n`

				rows, err := db.Query(context.Background(), sql)
				if err != nil {
					fmt.Println("Query err", err)
					continue
				}

				for rows.Next() {
					err := rows.Scan(nil, nil)
					if err != nil {
						fmt.Println("Scan err", err)
					}
				}

				if rows.Err() != nil {
					fmt.Println("rows.Err()", rows.Err())
				}
			}

			wg.Done()
		}()
	}

	wg.Wait()
	close(doneChan)
}
```

**Comment by jackc** [comment]
I just released `v4.8.1` which has a fix in place.

The problem was in the `pgconn` library. It would occur when an error was received between the `CommandComplete` message and the `ReadyForQuery` message. It was tricky to identify, but once it was I was able to make a test that mocked that specific sequence of messages that triggered the deadlock.

In this case `concludeCommand` would be called without an error. Then where the error was received it would be called again to store the error. But `concludeCommand` has a guard clause that only allows it to be run once. So it didn't store the error. This would lead to `(*ResultReader).Close` busy looping trying to read messages until it got `ReadyForQuery` but `receiveMessage` was returning a nil message *and* a nil error.

Fix: https://github.com/jackc/pgconn/commit/4e4c4ea5410aba437bc6d6e2c5a93c4acf6cce73
